File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  
ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
{
    ListNode *temp1 = l1, *temp2 = l2, *prev1 = nullptr, *prev2 = nullptr;
    int carry = 0;
    while (temp1 && temp2)
    {
        prev1 = temp1;
        prev2 = temp2;
        temp1->val = temp1->val + temp2->val + carry;
        temp2->val = temp1->val;
        carry = temp1->val / 10;
        if (temp1->val > 9)
        {
            carry = temp1->val / 10;
            temp1->val = temp1->val % 10;
            temp2->val = temp2->val % 10;
        }
        temp1 = temp1->next;
        temp2 = temp2->next;
    }
    if (temp1)
    {
        while (temp1)
        {
            prev1 = temp1;
            temp1->val += carry;
            carry = 0;
            if (temp1->val > 9)
            {
                carry = temp1->val / 10;
                temp1->val = temp1->val % 10;
            }
            temp1 = temp1->next;
        }
        if (carry)
        {
            ListNode *newNode = new ListNode(carry);
            prev1->next = newNode;
        }
        return l1;
    }
    else if (temp2)
    {
        while (temp2)
        {
            prev2=temp2;
            temp2->val += carry;
            carry = 0;
            if (temp2->val > 9)
            {
                carry = temp2->val / 10;
                temp2->val = temp2->val % 10;
            }
            temp2 = temp2->next;
        }
        if (carry)
        {
            ListNode *newNode = new ListNode(carry);
            prev2->next = newNode;
        }
       
        return l2;
    }
 if(carry)
        {
            ListNode *newNode = new ListNode(carry);
            prev1->next=newNode;
        }
    return l1;
}
};

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
 public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode temp1 = l1;
        ListNode temp2 = l2;
        ListNode ans = new ListNode(0);
        ListNode curr = ans;
        int carry = 0;

        while (temp1 != null && temp2 != null) {
            int val1 = temp1.val;
            int val2 = temp2.val;
            int sum = val1 + val2 + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode newNode = new ListNode(sum);
            curr.next = newNode;
            curr = curr.next;
            temp1 = temp1.next;
            temp2 = temp2.next;
        }
        while (temp1 != null) {
            int val1 = temp1.val;
            int sum = val1 + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode newNode = new ListNode(sum);
            curr.next = newNode;
            curr = curr.next;
            temp1 = temp1.next;
        }
        while (temp2 != null) {
            int val1 = temp2.val;
            int sum = val1 + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode newNode = new ListNode(sum);
            curr.next = newNode;
            curr = curr.next;
            temp2 = temp2.next;
        }
        if (carry != 0) {
            ListNode newNode = new ListNode(carry);
            carry = 0;
            curr.next = newNode;
            curr = curr.next;
        }
        return ans.next;
    }
}